### PR TITLE
Constrain a11y word nav to the virtual bottom

### LINKED ISSUE #5243: UIA: delay when reaching prompt with word navigation
# Environment

```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd] 1909 (18363.752)
Windows Terminal version (if applicable): Open Console (4f8acb4)

Any other software? NVDA, Narrator
```

# Steps to reproduce

1. Open Open Console.
2. With either NVDA or Narrator, move to the start of the console and begin navigation by word.
3. After reaching the word "reserved" (of the copyright text), move to the next word.

# Expected behaviour
The prompt is quickly read.

# Actual behaviour
A lag is observed before the prompt is read. A reduced lag (or none at all) is observed in Windows Terminal.

### PR DESCRIPTION
**Pull request by carlos-zamora**
Use the virtual bottom in ConHost to speed up a11y word navigation performance. `GetTextBufferEndPosition()` is also used by by search; it'll also get a minor performance boost.

## PR Checklist
* [X] Closes #5243 

## Detailed Description of the Pull Request / Additional comments
#7789 uses `GetLastNonspaceCharacter()` to determine where the last word is and exit early on a `UIA::Move` by word call. The change in this PR passes the virtual bottom over to `GetLastNonspaceCharacter()` to make that search a bit faster.

Tested word navigation on Host.exe using NVDA.

### WPR Traces
The percentages below represent the weight that a function call had. The
test scenario included moving by word on the CMD welcome message until
the last word was reached. Inspect.exe was used to limit any additional
calls that are generally performed by a screen reader.

| function   | current | branch |
| --         | --      | --     |
| `UIA:Move` | 29.52%  | 28.43% |

There is an improvement of about 1% in a release build of ConHost.

### CODE REVIEW

File: src/host/renderData.cpp [hunk 1]
```diff
@@ -33,9 +33,12 @@ Microsoft::Console::Types::Viewport RenderData::GetViewport() noexcept
...
 COORD RenderData::GetTextBufferEndPosition() const noexcept
 {
+    // We use the end line of mutableViewport as the end
+    // of the text buffer, it always moves with the written
+    // text
     const CONSOLE_INFORMATION& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-    Viewport bufferSize = gci.GetActiveOutputBuffer().GetBufferSize();
-    COORD endPosition{ bufferSize.Width() - 1, bufferSize.BottomInclusive() };
+    const auto mutableViewport = gci.GetActiveOutputBuffer().GetVirtualViewport();
+    COORD endPosition{ mutableViewport.Width() - 1, gsl::narrow<short>(mutableViewport.BottomInclusive()) };
```
Comment (zadjii-msft) [inline src/host/renderData.cpp:41]: I don't think you need the `gsl::narrow<short>` here

**Comment by codeofdusk** [comment]
Thanks @carlos-zamora!

File: src/host/renderData.cpp [hunk 2]
```diff
@@ -33,9 +33,12 @@ Microsoft::Console::Types::Viewport RenderData::GetViewport() noexcept
 // - COORD of the end position of the text buffer
 COORD RenderData::GetTextBufferEndPosition() const noexcept
 {
+    // We use the end line of mutableViewport as the end
+    // of the text buffer, it always moves with the written
+    // text
     const CONSOLE_INFORMATION& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-    Viewport bufferSize = gci.GetActiveOutputBuffer().GetBufferSize();
-    COORD endPosition{ bufferSize.Width() - 1, bufferSize.BottomInclusive() };
+    const auto mutableViewport = gci.GetActiveOutputBuffer().GetVirtualViewport();
```
Comment (DHowett-MSFT) [inline src/host/renderData.cpp:40]: it's not really _mutable_, but.. :P

File: src/host/renderData.cpp [hunk 3]
```diff
@@ -33,9 +33,12 @@ Microsoft::Console::Types::Viewport RenderData::GetViewport() noexcept
...
 COORD RenderData::GetTextBufferEndPosition() const noexcept
 {
+    // We use the end line of mutableViewport as the end
+    // of the text buffer, it always moves with the written
+    // text
     const CONSOLE_INFORMATION& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-    Viewport bufferSize = gci.GetActiveOutputBuffer().GetBufferSize();
-    COORD endPosition{ bufferSize.Width() - 1, bufferSize.BottomInclusive() };
+    const auto mutableViewport = gci.GetActiveOutputBuffer().GetVirtualViewport();
+    COORD endPosition{ mutableViewport.Width() - 1, mutableViewport.BottomInclusive() };
```
Comment (DHowett-MSFT) [inline src/host/renderData.cpp:41]: just so i understand: this is _not_ impacted by the user having scrolled up? the viewport _usually_ is.

**Comment by codeofdusk** [comment]
Any updates on this PR? Thanks.

**Comment by miniksa** [comment]
I came by to look at this but three things stick out:
1. Outstanding comments by @DHowett-MSFT that don't look answered yet.
2. The x64 build is failing.
3. @carlos-zamora hasn't attached the WPR traces before/after he said he was going to.

So I'm going to skip reviewing it for now and come back later hoping some of those progress.

**Comment by codeofdusk** [comment]
Do you intend to merge this in time for 21H1? It would mean a lot for NVDA. Thanks!

**Comment by carlos-zamora** [comment]
@codeofdusk Yeah! I've had bad luck the last few days where suddenly my internet goes down or my machine refuses to run the tracing app haha. I'm hoping to get this in this week or early next, but I'm trying to prioritize closing some of the other Windows Terminal bugs first.

**Comment by DHowett-MSFT** [comment]
@codeofdusk sorry about the confusion here: the 21H1 milestone end date was set incorrectly. Things are a lot less dire than “we need to fix all the UIA bugs in the next two days.”

**Comment by codeofdusk** [comment]
Excellent news!

**Comment by codeofdusk** [comment]
@carlos-zamora Any updates on this PR? Were you able to address the issues with your machine and outstanding review comments? Thanks!

**Comment by carlos-zamora** [comment]
> @carlos-zamora Any updates on this PR? Were you able to address the issues with your machine and outstanding review comments? Thanks!

Now that Build is over, time to get back to this one 😊. Getting the traces and fixing the build today.

**Comment by carlos-zamora** [comment]
> > @carlos-zamora Any updates on this PR? Were you able to address the issues with your machine and outstanding review comments? Thanks!
> 
> Now that Build is over, time to get back to this one 😊. Getting the traces and fixing the build today.

And another problem came up with recording the traces. Baby steps. This'll get done soon, sorry Bill haha.

**Comment by codeofdusk** [comment]
Thanks for looking into it!

Also, could you please look into #5481?

**Comment by DHowett** [comment]
@codeofdusk We'll look into it -- Carlos is assigned, but that does not indicate that it's in his immediate list of things to look at. We're finally coming off the v1 release cycle for Terminal and will have time in the future to evaluate conhost issues.

**Comment by codeofdusk** [comment]
What is the current status of this PR?

**Review by DHowett — APPROVED** [review]
This seems to have shaped up into something like we wanted it to. Can you change the name to something like "constrain a11y to the touched viewport"

consider that **SEARCH** uses this function (!) which could be problematic.

**Comment by codeofdusk** [comment]
Also, how does this PR relate to #6986?

**Comment by carlos-zamora** [comment]
> Also, how does this PR relate to #6986?

Here, we're making it easier to find the last non-space character (and the last word, by extension).

#6986 is an architectural change to treat the last non-space character as the end of the buffer. This change should make it a bit easier to accomplish that in a performant way.

**Comment by carlos-zamora** [comment]
Closing. See https://github.com/microsoft/terminal/issues/7000#issuecomment-809550014 for more details.

I'll revive this PR once I actively start working on UIA again.

**Pull request closed without merging**